### ISSUE TITLE
Need a JMH benchmark for virtual map reconnects

### ISSUE THREAD

**Issue by artemananiev** [hashgraph/hedera-services#11347]
We have a few unit tests (see `VirtualMapReconnectTest` and `MerkleTestUtils`) that simulate reconnects for very simple merkle trees, just a root node + a single virtual map. A new benchmark can be created off this code to test low-level virtual map reconnect performance. It won't be end-to-end (no network, no gossip, no incoming transactions, etc.), but it will still be very useful for debugging and performance assessments. Something similar to `VirtualMapBench` and `CryptoBench`, which are also low-level, yet very powerful.
```[tasklist]
### Tasks
- [ ] Add metrics support to the benchmark
```

**Comment by anthony-swirldslabs** [comment]
The benchmark introduced in #11487 uses in-memory data structures, and runs both the teacher and the learner in the same JVM. This means that whether the teacher needs to read the hash of a node only, or read the entire node and send it to the learner, these two operations would be largely identical as far as their overall performance is concerned. 

If we were to implement an optimization of the reconnect algorithm that tries to minimize the number of reads of entire nodes, then the current benchmarks wouldn't be able to detect any significant performance improvements (or regressions) because reading the entire node from memory is roughly as fast as only reading their hash.

From this perspective, measuring the time a reconnect takes when using in-memory data structures doesn't seem to be a correct approach. Further, even if we were to use an on-disk state generated for test purposes, then it would likely be relatively small to fit in the disk cache in memory, and hence it would produce similar performance effects (or absence thereof) on this benchmark as observed with in-memory data structures.

However, using a large, real state with many billions of nodes wouldn't allow us to host this benchmark in the repository as a local JMH benchmark because it would have to depend on retrieving this large test state which would negate the benefits of having a local JMH benchmark for quick performance measurements (it would be very slow to bootstrap and difficult/time-consuming to run locally.)

While adding hooks to the virtual map implementation and introducing artificial latencies for node retrieval operations might help in theory, this approach wouldn't be scientific (it's difficult to come up with reasonable latencies), and yet worse, this would artificially slow down the benchmark itself because it would have to waste CPU cycles on these added latencies.

Per https://javadoc.io/static/org.openjdk.jmh/jmh-core/0.5.6/org/openjdk/jmh/annotations/Mode.html , the JMH framework itself is only capable of measuring the time an operation takes, and per the above, it doesn't seem to be feasible to observe a time difference between different algorithms when running the benchmark in memory or with small test states. Therefore, it's unclear if JMH is the correct tool for measuring the performance of the reconnect algorithm improvements.

It would be more appropriate to measure specific metrics that we're trying to optimize in every particular case. For example, if an optimization aims to reduce the number of full nodes reads, then it is this counter that needs to be compared between the vanilla reconnect algorithm and the optimized version. While it's impossible to derive the exact time improvement that such an optimization would produce on a real reconnect run, it can be inferred (or at the very least, speculated) that reducing the amount of I/O operations should improve the overall performance of the reconnect process as a whole.

**Comment by anthony-swirldslabs** [comment]
The local benchmark introduced in #11487 appears to be useful as it revealed a few insights on the reconnect behavior. Specifically, I discovered that even a small number of missing nodes on the learner side changes the tree structure drastically which results in transferring pretty much all the leaves during the reconnect. Also, when all the nodes are present and only a small number of them have values different from the teacher's values, then the reconnect is indeed fast, with or without any optimizations. However, the latter case doesn't seem realistic, while the former case presents a challenge for naive reconnect optimizations. So I merged that PR.

At this point, I've started working on an integration-test-like reconnect benchmark which could use a local disk, and hence support larger states and most importantly, emulate the reconnect performance more precisely for the purpose of measuring it.
